### PR TITLE
feat: publish multi-arch images

### LINKED ISSUE #198: Publish images with arm64
Today, the bb-storage container image is published for x86_64 only.

```
% docker image inspect --format "{{.ID}} {{.Architecture}}" ghcr.io/buildbarn/bb-storage:20240311T070255Z-7b232a7
sha256:e1dcf8ca0b7b17a825ae8853aeb9db555b879302b851193592b239381eb32098 amd64
```

However there are some great instance types like https://instances.vantage.sh/aws/ec2/is4gen.medium which provide double the NVME per instance and 2.5X the I/O bandwidth and $0.144/hr vs $0.172 for `i4i.large`

These instances need arm64 architecture.
Note, I'm happy to work on a PR for this.

### PR DESCRIPTION
**Pull request by alexeagle**
Allows bb-storage users to deploy on arm64 architecture, e.g. AWS Graviton. Fixes #198

### CODE REVIEW

File: cmd/bb_copy/BUILD.bazel
Comment (alexeagle) [inline cmd/bb_copy/BUILD.bazel:36]: note, there's some code duplication between the three cmd packages. If you'd prefer I can add a macro in container.bzl like `multiarch_go_image`

**Comment by alexeagle** [comment]
I just learned about #167 , sorry this is replacing that I think

File: cmd/bb_copy/BUILD.bazel
Comment (EdSchouten) [reply cmd/bb_copy/BUILD.bazel:36]: I'd appreciate that. Otherwise we need to patch up this code every time we want to add support for another architecture.

File: cmd/bb_copy/BUILD.bazel
Comment (EdSchouten) [inline cmd/bb_copy/BUILD.bazel:34]: I don't think this change is acceptable for two reasons:

1. We no longer build Go binaries for the host architecture. We still want to have binaries for macOS, Windows, FreeBSD, etc..
2. `oci_image_index()` should just use transitions. That way there would be no need for us to put loops around `go_binary()`, `pkg_tar()`, and `oci_image()`.

File: tools/container.bzl
Comment (EdSchouten) [inline tools/container.bzl:8]: i suspect that you want to use `@com_github_buildbarn_bb_storage//tools:stamped_tags` here. Otherwise this won't be usable in bb-remote-execution and other repositories.

**Review by EdSchouten — CHANGES_REQUESTED** [review]

File: cmd/bb_replicator/BUILD.bazel
Comment (EdSchouten) [inline cmd/bb_replicator/BUILD.bazel:42]: Out of curiosity, why do we want to retain this path prefix?

Comment (alexeagle) [reply cmd/bb_replicator/BUILD.bazel:42]: I only did this to minimize the delta with the images you publish today. That's the path I see the binaries currently reside within the image.

Comment (EdSchouten) [reply cmd/bb_replicator/BUILD.bazel:42]: Feel free to remove it. As long as the resulting image is functionally equivalent, it's fine.

File: cmd/bb_copy/BUILD.bazel
Comment (alexeagle) [reply cmd/bb_copy/BUILD.bazel]: done, PTAL

File: tools/container.bzl
Comment (EdSchouten) [inline tools/container.bzl:18]: Considering that not all our executables are named "bb_storage", I think we should remove this.

Can't we simply use /app as the package dir?

Comment (alexeagle) [reply tools/container.bzl:18]: Sure, as long as the only reference to the path within the container is the entry_point. If users might have some config that has this path, it's obviously a breaking change to put the binary somewhere different.

**Comment by alexeagle** [comment]
Do you think my change is responsible for the gazelle failure on CI?

**Comment by EdSchouten** [comment]
> Do you think my change is responsible for the gazelle failure on CI?

Yes. I can reproduce it by running the commands run by the CI pipeline on my local system. It looks like I can address it by calling `gazell_dependencies()` earlier on:

```diff
diff --git a/WORKSPACE b/WORKSPACE
index 8806a54..9747956 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -73,6 +73,8 @@ go_rules_dependencies()
 
 go_register_toolchains(version = "1.21.5")
 
+gazelle_dependencies()
+
 load("@rules_oci//oci:pull.bzl", "oci_pull")
 load("@rules_oci//oci:repositories.bzl", "LATEST_CRANE_VERSION", "oci_register_toolchains")
 
@@ -101,8 +103,6 @@ oci_pull(
     ],
 )
 
-gazelle_dependencies()
-
 http_archive(
     name = "com_github_bazelbuild_buildtools",
     sha256 = "09a94213ea0d4a844e991374511fb0d44650e9c321799ec5d5dd28b250d82ca3",
```

File: tools/container.bzl
Comment (EdSchouten) [reply tools/container.bzl:18]: It should be. I don't think it's typical for people to override the entrypoint for the Buildbarn containers.

File: tools/container.bzl
Comment (EdSchouten) [inline tools/container.bzl:19]: Shouldn't this specify /app as a package dir?

File: tools/container.bzl
Comment (alexeagle) [reply tools/container.bzl]: okay, changed to just `app`

File: tools/container.bzl
Comment (EdSchouten) [inline tools/container.bzl:25]: Wouldn't this expand to /app/:bb_storage? Notice the colon.

**Comment by alexeagle** [comment]
PTAL - I have not integration-tested this by pushing a container and running a service from it.

File: tools/container.bzl
Comment (alexeagle) [reply tools/container.bzl:25]: yes, just added a removeprefix on that, I noticed as well

Comment (EdSchouten) [reply tools/container.bzl:25]: Can't we use Label(binary).name here instead?

File: tools/container.bzl
Comment (alexeagle) [reply tools/container.bzl:19]: yes, I pushed another commit right after this one, sorry you reviewed the intermediate state

File: tools/container.bzl
Comment (alexeagle) [reply tools/container.bzl:25]: not quite - in the macro context the string isn't fully qualified:

`Error in Label: Illegal absolute label syntax: invalid label ':bb_copy': absolute label must begin with '@' or '//'`

Comment (EdSchouten) [reply tools/container.bzl:25]: What about https://bazel.build/rules/lib/toplevel/native.html#package_relative_label ?

Comment (alexeagle) [reply tools/container.bzl:25]: Yeah I think that's what we want, but it was introduced in Bazel 6.1.0 https://github.com/bazelbuild/bazel/commit/b859571804f2095caaf018b172b59c90f185fd51 and this repo currently uses Bazel 6.0.0

Comment (EdSchouten) [reply tools/container.bzl:25]: Then let's try upgrading to 6.1.0 or later.

Comment (alexeagle) [reply tools/container.bzl:25]: pushed a commit to upgrade to Bazel 6.1.0, guessing that's what you'd prefer

File: WORKSPACE
Comment (stagnation) [inline WORKSPACE:15]: Speaking of bazel versions. I think you fix the blocker for bazel 7 here :)
https://github.com/buildbarn/bb-storage/issues/185

I never got around to verifying `rules_docker` builds on bazel 7 so thankful for this uplift.
We can look at bazel 7 after this PR :+1:

Comment (alexeagle) [reply WORKSPACE:15]: Yes that sounds correct.

**Comment by moroten** [comment]
I created https://github.com/bazel-contrib/rules_oci/pull/531 to make `oci_image_index` perform the target platform transition instead.

**Comment by EdSchouten** [comment]
I saw that that PR build of yours failed with some protoc code generation issues. I thought that would be caused by the Bazel 6.1.0 upgrade, so that's why I just did an upgrade to Bazel 6.4.0. Be sure to rebase your PR on top of `master` to see if it goes through now.

File: WORKSPACE
Comment (EdSchouten) [inline WORKSPACE:94]: Could you remove this comment and the NB below, please? We don't care strongly about using some exact version of the distroless image. I don't see the point in having some arbitrary version of an image documented.

**Comment by alexeagle** [comment]
@moroten FYI that I tried patching your rules_oci PR but it still fails on `linux/i386` in the same way as this PR does.

 https://github.com/buildbarn/bb-storage/actions/runs/8397776269/job/23006173303?pr=201

**Comment by alexeagle** [comment]
@EdSchouten I'm hoping my last commit made it green - do you have something like a label you could apply so the workflow doesn't require a manual approval when I push commits? 🙏🏻

**Comment by EdSchouten** [comment]
Github auto approves workflows if you have made at least 1 commit to this repo. So after I merge this change, any followup PR made by you should automatically run the build.

**Comment by alexeagle** [comment]
Hooray it's green, anything else you'd like me to change?

**Comment by moroten** [comment]
Thank you for merging this. The ideas with https://github.com/bazel-contrib/rules_oci/pull/531 for multi arch images will take some time to get ready.

**Pull request merged**
